CRM/Financial/BAO/Payment.php

```diff
@@ -56,62 +56,53 @@ public static function create($params) {
     $contributionStatus = CRM_Contribute_PseudoConstant::contributionStatus($contribution['contribution_status_id'], 'name');
 
-    // Check if pending contribution
-    $fullyPaidPayLater = FALSE;
-    if ($contributionStatus == 'Pending') {
-      $cmp = bccomp($contribution['total_amount'], $params['total_amount'], 5);
-      // Total payment amount is the whole amount paid against pending contribution
-      if ($cmp == 0 || $cmp == -1) {
-        civicrm_api3('Contribution', 'completetransaction', ['id' => $contribution['id']]);
-        // Get the trxn
-        $trxnId = CRM_Core_BAO_FinancialTrxn::getFinancialTrxnId($contribution['id'], 'DESC');
-        $ftParams = ['id' => $trxnId['financialTrxnId']];
-        $trxn = CRM_Core_BAO_FinancialTrxn::retrieve($ftParams, CRM_Core_DAO::$_nullArray);
-        $fullyPaidPayLater = TRUE;
-      }
-      else {
-        civicrm_api3('Contribution', 'create',
-          [
-            'id' => $contribution['id'],
-            'contribution_status_id' => 'Partially paid',
-          ]
-        );
-      }
+    $cmp = bccomp($contribution['total_amount'], $params['total_amount'], 5);
+    $isPaymentCompletesContribution = ($cmp == 0 || $cmp == -1);
+
+    if ($contributionStatus == 'Pending' && !$isPaymentCompletesContribution) {
+      civicrm_api3('Contribution', 'create',
+        [
+          'id' => $contribution['id'],
+          'contribution_status_id' => 'Partially paid',
+        ]
+      );
     }
-    if (!$fullyPaidPayLater) {
-      $trxn = CRM_Core_BAO_FinancialTrxn::getPartialPaymentTrxn($contribution, $params);
-      if (CRM_Utils_Array::value('line_item', $params) && !empty($trxn)) {
-        foreach ($params['line_item'] as $values) {
-          foreach ($values as $id => $amount) {
-            $p = ['id' => $id];
-            $check = CRM_Price_BAO_LineItem::retrieve($p, $defaults);
-            if (empty($check)) {
-              throw new API_Exception('Please specify a valid Line Item.');
-            }
-            // get financial item
-            $sql = "SELECT fi.id
-            FROM civicrm_financial_item fi
-            INNER JOIN civicrm_line_item li ON li.id = fi.entity_id and fi.entity_table = 'civicrm_line_item'
-            WHERE li.contribution_id = %1 AND li.id = %2";
-            $sqlParams = [
-              1 => [$params['contribution_id'], 'Integer'],
-              2 => [$id, 'Integer'],
-            ];
-            $fid = CRM_Core_DAO::singleValueQuery($sql, $sqlParams);
-            // Record Entity Financial Trxn
-            $eftParams = [
-              'entity_table' => 'civicrm_financial_item',
-              'financial_trxn_id' => $trxn->id,
-              'amount' => $amount,
-              'entity_id' => $fid,
-            ];
-            civicrm_api3('EntityFinancialTrxn', 'create', $eftParams);
+
+    $trxn = CRM_Contribute_BAO_Contribution::recordPartialPayment($contribution, $params);
```

Comment:
This doesn't handle the refund case, so I was thinking to have : 
```
// pseudocode
if ($refund) {
$trxn = recordRefundPayment();
// OR should call payment.refund API because I am working on a feature which post refund request via PP
}
else {
$trxn = recordPartialPayment();
}

Reply:
@monishdeb yeah I left that out of scope but TBH I think we should just treat any negative payment as a refund & call the stuff that is in recordAdditionalPayment

I think the distinction between Payment & Refund is not important here but when we have PaymentProcessor.pay & PaymentProcessor.refund then it matters

Reply:
in that for now should we move the refund call to ```recordAdditionalPayment()``` on basis of status? But I think we should rather move it in a separate BAO fn for clarity and then drop ```recordAdditionalPayment()``` definition

Reply:
once we have gotten payment.cancel to call payment.create - per the other PR & the additionalPayment form to call Payment.create recordAdditionalPayment will be **obsolete** - yay! 

Towards getting there let's maybe we should FIRST extract the refund handling part to it's own function that we can call from here & from recordAdditionalPayment

Having said that - I think THIS PR can leave refund out of scope & that can be NEXT PR -

Reply:
Maybe its just me :) that I don't like the API name to be Payment.cancel  that handles refund. But I would like the idea to call payment.cancel in case when contribution_status = Cancelled.

Reply:
Yeah - I think Payment.refund would have been a better api name - we could create an alias for it. But for the current scope - it already exists & is somewhat tested

I don't know what you would do to payments if the contribution is cancelled - that feels very internal rather than api-y